File: app/src/main/java/com/carbonylgroup/schoolpower/activities/LoginActivity.java
/**
 * Copyright (C) 2017 Gustav Wang
 */

package com.carbonylgroup.schoolpower.activities;


import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.design.widget.Snackbar;
import android.view.View;
import android.widget.EditText;

import java.security.PublicKey;

import com.carbonylgroup.schoolpower.R;
import com.carbonylgroup.schoolpower.classes.Utils.Utils;
import com.carbonylgroup.schoolpower.classes.Utils.postData;

public class LoginActivity extends Activity {

    private Utils utils;

    private final static String publicKeyString = "" +
            "MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDtILvI2l7yH/9yn/qKmpbqGqSi" +
            "sawP42X1Js4zykHgGdqSyQ4PsKEbEEWh8KLOdaCeBkxMRzqhKS3WLI78oKijNgg7" +
            "6z0/jHJoCrOEJmCeWA2ugcgUOrw5i2siFHo/ogHQhtCf0fa1a+6PUjwOvhhaU4yW" +
            "EXfsfWVY0iJA4qO58wIDAQAB";

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        setTheme(R.style.Theme_Panzer);
        super.onCreate(savedInstanceState);
        checkIfLoggedIn();
        setContentView(R.layout.login_content);

        initValue();
    }

    private void checkIfLoggedIn() {

        SharedPreferences sharedPreferences = getSharedPreferences(getString(R.string.accountData), Activity.MODE_PRIVATE);
        if (sharedPreferences.getBoolean(getString(R.string.loggedIn), false))
            startMainActivity();
    }

    private void initValue() {

        utils = new Utils(this);

        final EditText input_username = (EditText) findViewById(R.id.input_username);
        final EditText input_password = (EditText) findViewById(R.id.input_password);

        findViewById(R.id.login_fab).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                loginAction(input_username.getText().toString(), input_password.getText().toString());
            }
        });
    }

    public void loginAction(final String username, final String password) {

        final ProgressDialog progressDialog = new ProgressDialog(LoginActivity.this);
        progressDialog.setIndeterminate(true);
        progressDialog.setCancelable(false);
        progressDialog.setCanceledOnTouchOutside(false);
        progressDialog.setMessage(getString(R.string.authenticating));
        progressDialog.show();

        PublicKey publicKey = Utils.restorePublicKey(publicKeyString);

        final String encryptedArgument = Utils.RSAEncode(publicKey, username + ";" + password);

        new Thread(new postData(
                getString(R.string.postURL),
                getString(R.string.token_equals) + encryptedArgument + "&filter=",
                new Handler() {
                    @Override
                    public void handleMessage(Message msg) {

                        progressDialog.dismiss();
                        String message = msg.obj.toString();
                        if (message.contains(getString(R.string.error_wrong_password)))
                            showSnackBar(getString(R.string.wrong_password), true);

                        else if (message.contains("[]")) {
                            SharedPreferences.Editor spEditor = getSharedPreferences(getString(R.string.accountData), Activity.MODE_PRIVATE).edit();
                            spEditor.putString(getString(R.string.token), encryptedArgument);
                            spEditor.putBoolean(getString(R.string.loggedIn), true);
                            spEditor.apply();
                            try {
                                utils.saveDataJson(message);
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                            startMainActivity();

                        } else
                            showSnackBar(getString(R.string.no_connection), true);
                    }
                })).start();

    }

    private void startMainActivity() {

        startActivity(new Intent(getApplication(), MainActivity.class));
        LoginActivity.this.finish();
    }

    private void showSnackBar(String msg, boolean colorRed) {

        Snackbar snackbar = Snackbar.make(findViewById(R.id.login_coordinate_layout), msg, Snackbar.LENGTH_SHORT);
        if (colorRed)
            snackbar.getView().setBackgroundColor(getResources().getColor(R.color.Cm_score_red_dark));
        else snackbar.getView().setBackgroundColor(getResources().getColor(R.color.accent));
        snackbar.show();
    }
}

File: app/src/main/java/com/carbonylgroup/schoolpower/activities/LoginActivity.kt
/**
 * Copyright (C) 2019 SchoolPower Studio
 */

package com.carbonylgroup.schoolpower.activities

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import androidx.core.content.ContextCompat
import androidx.appcompat.widget.AppCompatTextView
import android.util.Log
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.ProgressBar
import com.carbonylgroup.schoolpower.R
import com.carbonylgroup.schoolpower.data.StudentData
import com.carbonylgroup.schoolpower.utils.Utils
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MultipartBody
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException


class LoginActivity : BaseActivity() {

    override fun initActivity() {
        super.initActivity()
        setContentView(R.layout.login_content)
        window.navigationBarColor = ContextCompat.getColor(this, R.color.primary)
        window.statusBarColor = ContextCompat.getColor(this, R.color.primary)

        hideProgress()
        initDialog()

        findViewById<View>(R.id.login_fab).setOnClickListener {
            val username = findViewById<EditText>(R.id.input_username).text.toString()
            val password = findViewById<EditText>(R.id.input_password).text.toString()
            utils.getPreferences(Utils.AccountData).edit()
                    .putString(getString(R.string.user_id), username)
                    .apply()
            if (username != "")
                loginAction(username, password)
            else
                utils.showSnackBar(findViewById(R.id.login_coordinate_layout), getString(R.string.wrong_password), true)

        }
    }

    private fun initDialog() {
        val builder = AlertDialog.Builder(this)
        builder.setMessage(getString(R.string.only_alert))
        builder.setTitle(getString(R.string.notification))
        builder.setPositiveButton(getString(R.string.i_understand), null)
        builder.create().show()
    }

    private fun showProgress() {
        runOnUiThread {
            findViewById<ProgressBar>(R.id.login_progress_bar).visibility = View.VISIBLE
            findViewById<LinearLayout>(R.id.login_scroll).visibility = View.INVISIBLE
            findViewById<AppCompatTextView>(R.id.login_copyright).visibility = View.INVISIBLE
        }
    }

    private fun hideProgress() {
        runOnUiThread {
            findViewById<ProgressBar>(R.id.login_progress_bar).visibility = View.GONE
            findViewById<LinearLayout>(R.id.login_scroll).visibility = View.VISIBLE
            findViewById<AppCompatTextView>(R.id.login_copyright).visibility = View.VISIBLE
        }
    }

    private fun hideSoftKeyboard() {
        val view = this.currentFocus
        if (view != null) {
            val imm = this.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    private fun loginAction(username: String, password: String) {

        hideSoftKeyboard()
        showProgress()

        val body = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("username", username)
                .addFormDataPart("password", password)
                .addFormDataPart("version", utils.getAppVersion())
                .addFormDataPart("action", "login")
                .addFormDataPart("os", "android")
                .build()

        var retried = false
        utils.buildNetworkRequest(getString(R.string.postURL), "POST",
                body).enqueue(
                object : Callback {
                    override fun onFailure(call: Call, e: IOException) {
                        e.printStackTrace()
                        val backupServer = utils.getBackupServerUrl("pull_data_2")
                        if (!retried && backupServer != null) {
                            retried = true
                            try {
                                val response = utils.buildNetworkRequest(backupServer, "POST", body).execute()
                                onResponse(call, response)
                            } catch (e: IOException) {
                            }
                            return
                        }
                        hideProgress()
                        utils.showSnackBar(findViewById(R.id.login_coordinate_layout), getString(R.string.no_connection), true)
                    }

                    override fun onResponse(call: Call, response: Response) {

                        val strMessage = response.body()!!.string().replace("\n", "")
                        response.close()
                        // Wrong username/password
                        if (strMessage.contains("Invalid Username or password")) {
                            runOnUiThread {
                                utils.showSnackBar(findViewById(R.id.login_coordinate_layout), getString(R.string.wrong_password), true)
                                Log.w("Login", strMessage)
                                hideProgress()
                            }
                            return
                        }
                        if (strMessage.contains("\"alert\"")) {
                            runOnUiThread {
                                utils.showSnackBar(findViewById(R.id.login_coordinate_layout), JSONObject(strMessage)["alert"].toString(), true)
                                hideProgress()
                            }
                            return
                        }
                        if (!strMessage.contains("{")) {
                            runOnUiThread {
                                utils.showSnackBar(findViewById(R.id.login_coordinate_layout), getString(R.string.no_connection), true)
                                hideProgress()
                            }
                            return
                        }

                        val data =
                                try { StudentData(this@LoginActivity, strMessage, utils) }
                                catch (e: IllegalArgumentException){
                                    if(e.message?.contains("ERROR_ACCOUNT_DISABLED") == true) {
                                        runOnUiThread {
                                            AlertDialog.Builder(this@LoginActivity)
                                                    .setMessage(getString(R.string.account_disabled))
                                                    .setTitle(getString(R.string.account_disabled_desc))
                                                    .setPositiveButton(getString(R.string.alright), null)
                                                    .create().show()
                                        }
                                    } else {
                                        utils.errorHandler(e, getString(R.string.fatel_error_server_side),
                                                getString(R.string.fatel_error_server_side_message) + e.message)
                                    }
                                    runOnUiThread {
                                        hideProgress()
                                    }
                                    null
                                } ?: return
                        utils.getPreferences(Utils.AccountData).edit()
                                .putString(getString(R.string.usernameKEY), username)
                                .putString(getString(R.string.passwordKEY), password)
                                .putBoolean(getString(R.string.loggedIn), true)
                                .putString(getString(R.string.student_name), data.studentInfo.getFullName())
                                .apply()

                        utils.saveDataJson(strMessage)
                        utils.saveHistoryGrade(data.subjects)

                        startMainActivity()
                    }
                })
    }

    private fun startMainActivity() {
        startActivity(Intent(application, MainActivity::class.java))
        this@LoginActivity.finish()
    }
}
